src/types/weakref/reference.rs

```diff
@@ -0,0 +1,1200 @@
+use crate::err::PyResult;
+use crate::ffi_ptr_ext::FfiPtrExt;
+use crate::type_object::{PyTypeCheck, PyTypeInfo};
+use crate::types::any::PyAnyMethods;
+use crate::{ffi, Borrowed, Bound, PyAny, PyNativeType, Python, ToPyObject};
+
+/// Represents a Python `weakref.ReferenceType`.
+///
+/// In Python this is created by calling `weakref.ref`.
+#[repr(transparent)]
+pub struct PyWeakRef(PyAny);
+
+pyobject_native_type!(
+    PyWeakRef,
+    ffi::PyWeakReference,
+    pyobject_native_static_type_object!(ffi::_PyWeakref_RefType),
+    #module=Some("weakref"),
+    #checkfunction=ffi::PyWeakref_CheckRefExact
+);
+
+impl PyWeakRef {
+    /// Deprecated form of [`PyWeakRef::new_bound`].
+    #[inline]
+    #[track_caller]
+    #[cfg_attr(
+        not(feature = "gil-refs"),
+        deprecated(
+            since = "0.21.0",
+            note = "`PyWeakRef::new` will be replaced by `PyWeakRef::new_bound` in a future PyO3 version"
+        )
+    )]
+    pub fn new<T>(py: Python<'_>, object: T) -> PyResult<&'_ PyWeakRef>
+    where
+        T: ToPyObject,
+    {
+        Self::new_bound(py, object).map(Bound::into_gil_ref)
+    }
```

Comment:
It was very diligent of you to implement all of these deprecations etc. I think it probably offers the best experience for users, so thank you 🙏